detect/utils.go

```diff
@@ -7,34 +7,67 @@ import (
 	"strings"
 	"time"
 
-	"github.com/charmbracelet/lipgloss"
-
+	"github.com/zricethezav/gitleaks/v8/cmd/scm"
 	"github.com/zricethezav/gitleaks/v8/logging"
 	"github.com/zricethezav/gitleaks/v8/report"
 
+	"github.com/charmbracelet/lipgloss"
 	"github.com/gitleaks/go-gitdiff/gitdiff"
 )
 
 // augmentGitFinding updates the start and end line numbers of a finding to include the
 // delta from the git diff
-func augmentGitFinding(finding report.Finding, textFragment *gitdiff.TextFragment, f *gitdiff.File) report.Finding {
+func augmentGitFinding(scmProvider scm.Provider, remoteUrl string, finding report.Finding, textFragment *gitdiff.TextFragment, f *gitdiff.File) report.Finding {
 	if !strings.HasPrefix(finding.Match, "file detected") {
 		finding.StartLine += int(textFragment.NewPosition)
 		finding.EndLine += int(textFragment.NewPosition)
 	}
 
 	if f.PatchHeader != nil {
 		finding.Commit = f.PatchHeader.SHA
-		finding.Message = f.PatchHeader.Message()
 		if f.PatchHeader.Author != nil {
 			finding.Author = f.PatchHeader.Author.Name
 			finding.Email = f.PatchHeader.Author.Email
 		}
 		finding.Date = f.PatchHeader.AuthorDate.UTC().Format(time.RFC3339)
+		finding.Message = f.PatchHeader.Message()
+		// Results from `git diff` shouldn't have a link.
+		if finding.Commit != "" {
+			finding.Link = createScmLink(scmProvider, remoteUrl, finding)
+		}
 	}
 	return finding
 }
 
+func createScmLink(scmProvider scm.Provider, remoteUrl string, finding report.Finding) string {
+	if scmProvider == scm.NoProvider {
+		return ""
+	}
+
+	// Clean the path.
+	filePath := strings.Replace(finding.File, "%s", "%25", -1)
+
+	switch scmProvider {
+	case scm.GitHubProvider:
+		link := fmt.Sprintf("%s/blob/%s/%s", remoteUrl, finding.Commit, filePath)
+		if strings.HasSuffix(link, ".md") {
+			link += "?plain=1"
+		}
+		if finding.StartLine != 0 {
+			link += fmt.Sprintf("#L%d", finding.StartLine)
+		}
+		if finding.EndLine != finding.StartLine {
+			link += fmt.Sprintf("-L%d", finding.EndLine)
+		}
+		return link
+	case scm.GitLabProvider:
+		return fmt.Sprintf("%s/blob/%s/%s#L%d", remoteUrl, finding.Commit, filePath, finding.StartLine)
```

Comment:
@rgmz does GitLab not support multi-line links?

Reply:
Good catch. I had a few unsuccessful attempts to select multiple lines, I guess it's finicky.

Reply:
Done.